pkg/status/nodestatus.go

```diff
@@ -0,0 +1,300 @@
+// - If we get a value update, cache it to pendingUpdates.
+// - If we get a inSync message, set inSync to true.
+//   We don't need to worry about any status message after inSync message,
+//   getting a non-in sync status after an in-sync isn't important here, there's no real impact.
+//   It'd just mean that we've got slightly old data.
+// - After handling syncer event, process pendingUpdates if we are in-sync.
+func (r *nodeStatusReporter) run() {
+	// Loop forever, updating whenever we get a kick. The first kick will happen as soon as the syncer is in sync.
+	for {
+		select {
+		case e := <-r.syncerC:
+			switch event := e.(type) {
+			case []bapi.Update:
+				r.onUpdates(event)
+			case bapi.SyncStatus:
+				if event == bapi.InSync {
+					r.inSync = true
+				}
+			default:
+				log.Panicf("Unknown type %T in syncer channel", event)
+			}
+		case <-r.done:
+			return
+		}
+
+		if r.inSync {
+			r.processPendingUpdates()
+		}
+	}
+}
+
+// OnUpdated handles the syncer update callback method.
+func (r *nodeStatusReporter) OnUpdates(updates []bapi.Update) {
+	r.syncerC <- updates
+}
+
+// OnStatusUpdated handles the syncer status callback method.
+func (r *nodeStatusReporter) OnStatusUpdated(status bapi.SyncStatus) {
+	if status == bapi.InSync {
+		r.syncerC <- status
+	}
+}
+
+// onUpdates caches the syncer resource updates in main loop.
+func (r *nodeStatusReporter) onUpdates(updates []bapi.Update) {
+	for _, u := range updates {
+		var name string
+		// Get resource name from the key.
+		// Node status is non-namespaced resources hence
+		// resource name is unique.
+		if v, ok := u.Key.(model.ResourceKey); ok {
+			name = v.Name
+		} else {
+			log.Warningf("Unexpected resource update: %s", u.Key)
+			continue
+		}
+
+		switch u.UpdateType {
+		case bapi.UpdateTypeKVDeleted:
+			// Resource is deleted. Set nil pointer for pending updates.
+			r.pendingUpdates[name] = nil
+			log.Infof("Deleted node status resource: %s", u.Key)
+		case bapi.UpdateTypeKVNew, bapi.UpdateTypeKVUpdated:
+			if u.Value == nil {
+				// Value could be nil if typha failed to validate the resource.
+				log.Debugf("Nil value on new or updated KV: %s", u.Key)
+				continue
+			}
+			// Resource is created or updated. Cache latest value to pending updates.
+			switch v := u.Value.(type) {
+			case *apiv3.CalicoNodeStatus:
+				if v.Spec.Node != r.nodename {
+					// Node status request is not for us.
+					continue
+				}
+				if v.Spec.UpdatePeriodSeconds == nil {
+					log.Errorf("UpdatePeriodSeconds not set for node status resource: %s", u.Key)
+					continue
+				}
+				log.Infof("Updated node status resource: %s", u.Key)
+				r.pendingUpdates[name] = v
+			default:
+				log.Warningf("Unexpected resource update: %s", u.Key)
+				continue
+			}
+		}
+	}
+}
+
+// processPendingUpdates processes pending updates in main loop.
+// It is called when we are in-sync.
+func (r *nodeStatusReporter) processPendingUpdates() {
+	if len(r.pendingUpdates) == 0 {
```

Comment:
This is redundant code - if pendingUpdates is empty, the loop below will just be skipped.